### PR TITLE
Autotools packages: modify ACLOCAL_PATH to build dependent packages

### PR DESCRIPTION
**Pull request by tkameyama**
closes #19258 

gettext has m4 files for aclocal (for example, gettext.m4, iconv.m4).
But gettext package dose not set ACLOCAL_PATH.
This PR set ACLOCAL_PATH.

### CODE REVIEW

**Comment by adamjstewart** [comment]
It might make more sense to put this in `etc/spack/defaults/modules.yaml` so it is set for all packages that have that directory. We do something similar in the `AutotoolsPackage.autoreconf` method where we search for that directory in all dependency packages.

File: lib/spack/spack/build_systems/autotools.py
Comment (alalazo) [inline lib/spack/spack/build_systems/autotools.py:250]: Shouldn't this go in `setup_build_environment`?

Comment (tkameyama) [reply lib/spack/spack/build_systems/autotools.py:250]: I  want to avoid replacing `setup_build_environment` with the package,py

Comment (alalazo) [reply lib/spack/spack/build_systems/autotools.py:250]: @tkameyama Ugh, forget my comment above. I saw the title of the PR starting with `gettext` and didn't notice this change was not referring to that package but to the `autotools` build system.

**Comment by tkameyama** [comment]
> 
> 
> It might make more sense to put this in `etc/spack/defaults/modules.yaml` so it is set for all packages that have that directory. We do something similar in the `AutotoolsPackage.autoreconf` method where we search for that directory in all dependency packages.

The package replace autoreconf (For example audacious) dose not search that directory.
So I change in Autoreconf package.

**Review by alalazo — APPROVED** [review]
I agree with setting `ACLOCAL_PATH` for build dependencies that have `share/aclocal` directories.

@adamjstewart Python is dynamically typed and I think I might be able to use meta-classes to do the same thing done here as part of our `setup_*_environment` protocol without changing any package code. Let me know if that would be preferable (since it should be reflected into modules as well) and I'll give it a try.

File: lib/spack/spack/build_systems/autotools.py
Comment (alalazo) [reply lib/spack/spack/build_systems/autotools.py:250]: After thinking a bit about that, I believe I can employ metaclasses to have this bit of code in `setup_*_environment` without changing any `package.py` implementation.

**Comment by alalazo** [comment]
@tkameyama I had a better look at the context, and it seems we already have something similar (as @adamjstewart pointed out) in `AutotoolsPackage.autoreconf`:

https://github.com/spack/spack/blob/8edb31e934e38a7f0f1b6c7a6da42ac1a7dd9591/lib/spack/spack/build_systems/autotools.py#L264-L270

That change was introduced in #15784 with the only difference that it's using the `-I` option to extend the macro search paths instead of the `ACLOCAL_PATH` environment variable. Before merging the PR I want to double check with you why this duplication is needed. Is there some reproducible use case where the `-I` approach doesn't work but `ACLOCAL_PATH` does?

**Comment by tkameyama** [comment]
> 
> 
> @tkameyama I had a better look at the context, and it seems we already have something similar (as @adamjstewart pointed out) in `AutotoolsPackage.autoreconf`:
> 
> https://github.com/spack/spack/blob/8edb31e934e38a7f0f1b6c7a6da42ac1a7dd9591/lib/spack/spack/build_systems/autotools.py#L264-L270
> 
> That change was introduced in #15784 with the only difference that it's using the `-I` option to extend the macro search paths instead of the `ACLOCAL_PATH` environment variable. Before merging the PR I want to double check with you why this duplication is needed. Is there some reproducible use case where the `-I` approach doesn't work but `ACLOCAL_PATH` does?

Please check audacious (sorry you may need fix #18884)
audacious cannot use this function, because audacious use own autoreconf.

File: lib/spack/spack/build_systems/autotools.py
Comment (alalazo) [reply lib/spack/spack/build_systems/autotools.py:250]: For the record, the approach I propose above can be done by intercepting class creation and patching `setup_*_environment` to add the autotools specific modifications. Said otherwise, with meta-classes we can decorate any method on the fly and programmatically. Not sure it's worth it though, since it's more code than this and more cryptic.

**Comment by alalazo** [comment]
@adamjstewart @tkameyama I went back looking at this PR. The code looks good to me, hence the approval. Nonetheless I think we should either:
1. Merge this PR after having removed the code shown in https://github.com/spack/spack/pull/18582#issuecomment-700606433 that set the `-I` search paths (which means "switch to setting ACLOCAL_PATH for every autotools package")
2. Or refactor that code as shown in #19258 and continue to use only `-I` In that PR I used `audacious` as an example since it was mentioned as a problematic package

Said otherwise the only thing that kept me from merging this PR is that we would have two competing mechanisms to do the same thing, which need to be kept in sync. Thoughts?

**Comment by adamjstewart** [comment]
I like #19258

**Pull request closed without merging**
